### ISSUE TITLE
Make CCTP and Community Patch compatible plus update external mods

### ISSUE THREAD

**Issue by WSLUser** [LoneGazebo/Community-Patch-DLL#5496]
_1. Mod version (i.e Date - 4/23):_
v89 4/29 per .modinfo

_2. Mod list (if using Vox Populi only, leave blank):_
Too large to put here but basically:

Community Patch
Community Call to Power
Bunch of mods from http://www.picknmixmods.com/main/home/home.html
Some various building/wonder mods that never had issue when using Community Patch or CCTP separetely
WonderRace
BeyondtheFuture
GameOptions
Ancients civ and other 2 from same author
_3. Error description:_ 
CTD immediately when attempting to start game. Investigating, I've discovered Community Patch is using some mods from http://www.picknmixmods.com/main/home/home.html that CCTP also uses. SmartAI being one them. Community Patch used to be compatible so if those were always there, then something else is going on but pretty sure using duplicates will cause some breakage in loading or later ingame.

_4. Steps to reproduce (optional):_

Load DLL - Various Mod Components
Load CP and CCTP. Configure a game. Optional steps: I used legendary start for resources and specified all civs. Doesn't matter which civs are specified. Use cruel option for game speed. Also maxed out city states.
---------------------------
Supporting information:
Please note that you can attach .zip files by dragging-and-dropping them. If possible, zip up all supporting data and post that way.

1. Log files (always attach your Logs folder, located at My Documents/My Games/Sid Meier's Civilization 5. Make sure you have enabled logging before experiencing an error! Go here to find out how: http://forums.civfanatics.com/showthread.php?t=487482): Will upload tomorrow. Need to generate.

2. Save game (always attach a save that was made a turn before the error; located at My Documents/My Games/Sid Meier's Civilization 5/ModdedSaves): N/A.

3. CvMiniDump.dmp file (attach if experiencing a game crash. Located at .../Steam/steamapps/common/Sid Meier's Civilization V): Wasn't there.

4. Screenshots (optional):

**Comment by WSLUser** [comment]
Ideally CBP could be used as well but that would take far more significant resources. But at it's last developed stated, CCTP wasn't completely balanced.

**Comment by MoiMagnus** [comment]
From my limited knowledge, CP isn't compatible with "DLL - Various Mod Compotnens", as you can never have two DLL mods together (and CP is a DLL mod).

So, for CP and CCTP to be compatible, one should ensure that CP contain everything "DLL - Various Mod Components" contains. It contains already a significant part of it, but I'm not sure it contains everything, and CP made some **major** changes code-wise in the last 2 years, which might cause some compatibility problems.

Have you tried to play with CCTP and CP without "DLL - Various Mod Compotnens"?

**Comment by Vyyt02** [comment]
Yes, what MoiMagnus said. You cannot have 2 DLLs active at the same time. CP already includes the DLL various components and SmartAI, so try without these 2.

**Comment by Vyyt02** [comment]
Actually try just the 2 first (CP and the CCTP), no other mods, then you will know where to start looking for the incompatibility.

**Comment by WSLUser** [comment]
Ah I forgot that CP already included DLL - Various Mod Components. CCTP just needs that to load. If CP already has it, then CCTP won't care if it uses the one provided from CP. Strange though, because I thought I saw CP say it needed it loaded before I could activate it. Also, I found the dump file. Tried reading it but the encoding isn't friendly for viewing. Here it is though if you know how to get it open. 
[CvMiniDump.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/3390289/CvMiniDump.zip)

**Comment by WSLUser** [comment]
That still won't address the duplicate mods though. It may work but it would be more stable if CP dropped the mods that CCTP uses. Not as the main CP of course but as an alternative. Also, as I noticed some dependency mods are out of date and I think they should be updated/upgraded to their latest versions. Especially since those latest versions are still several years old. For starters the "Smart AI" I already mentioned has a V3 that neither CCTP or CP uses. I would of fixed updating these mods myself but not sure how the md5=7878789rfy7sdy for hashing is being done in .modinfo. Is that something done in ModBuddy? I'm not familiar with using that IDE (that really needs an update but never will) but if it's generated from there, a quick pointer to using it would be appreciated.

**Comment by MoiMagnus** [comment]
> CCTP just needs that to load.

In the .modinfo file of CCTP, there is a line just after "<Dependencies>" that states that VMC is needed. Replace it by:

> <Mod id="d1b6328c-ff44-4b0d-aad7-c657f83610cd" minversion="0" maxversion="999" title="Community Patch" />

So that it will require Community Patch instead of VMC

**Comment by Vyyt02** [comment]
I think CP is just blocking both the original Smart AI mod and the WHoward's updated version in order to prevent the conflicts between the mods. Trust me CP is the best AI mod for Civ5.

**Comment by MoiMagnus** [comment]
> That still won't address the duplicate mods though. It may work but it would be more stable if CP dropped the mods that CCTP uses. Not as the main CP of course but as an alternative. Also, as I noticed some dependency mods are out of date and I think they should be updated/upgraded to their latest versions. Especially since those latest versions are still several years old. For starters the "Smart AI" I already mentioned has a V3 that neither CCTP or CP uses. I would of fixed updating these mods myself but not sure how the md5=7878789rfy7sdy for hashing is being done in .modinfo. Is that something done in ModBuddy? I'm not familiar with using that IDE (that really needs an update but never will) but if it's generated from there, a quick pointer to using it would be appreciated.

Whatever CP would drop, it would **never** be compatible with "DLL - Various Mod Components". You can't have two DLL mods. Even if you enable both, only one will work (and the other one will just cause errors). There is a unique .dll file used by the game at once. If you don't have any DLL mod, it uses the default BNW dll file. If you have DLL mods, it uses the .dll file of the last DLL mod loaded.

When you have two DLL mods, the only way to make them compatible is to merge the two into a single mod (which can be a long and tedious procedure). To edit DLL files, you have to obtain the sources C++ files (that you can find on the Github), edit them, and recompile them into a .dll file that you drop in your mod. ModBuddy is only useful for .xml, .sql and .lua, but can't make .dll files.

CP didn't upgrade to Smart AI v3 because there is no point: every part of the AI is overwritten by CP in stuff that are hopefully more efficient than Smart AI (v2 or v3). Any tentative to mix the CP AI with another mod that alter the AI will likely just make the AI worse (because of compatibility bugs).

**Comment by WSLUser** [comment]
> ModBuddy is only useful for .xml, .sql and .lua, but can't make .dll files.

At this point, I'd say VS Code is superior. There are extensions available (xml and sql from MS), lua from others.  Needless to say the actual latest Visual Studio IDE also would support those better. 

I made the change so now I'm going to test it. If it works, then I'll worry about the dupes if something shows up.

**Comment by WSLUser** [comment]
I got it working with just those 2 mods but had to clean re-install. It had actually gotten worse to where I couldn't even get to the setup screen and that's after fixing an issue with CP deciding to disappear and couldn't be brought back.  I verified I could get to turn 2. Now I have the unenviable job of determining what other mods I can add back. Having a full list of everything CP uses would go a great way for helping that to avoid duplicates.
